extensions/amp-a4a/0.1/test/test-amp-a4a.js

```diff
@@ -124,6 +139,198 @@ describe('amp-a4a', () => {
+          a4a.vsync_.runScheduledTasks_();
+	  expect(getAdUrlSpy.calledOnce,
+		 'getAdUrlSpy called exactly once').to.be.true;
+	  expect(getAdUrlSpy.exceptions.length).to.equal(1);
+	});
+      });
+    });
+  });
+
+  describe('#sendXhrRequest_', () => {
+    it('should be called once', () => {
+      viewerWhenVisibleMock.onFirstCall().returns(Promise.resolve());
+      xhrMock.withArgs('https://test.location.org/ad/012345?args', {
+        mode: 'cors',
+        method: 'GET',
+        credentials: 'include',
+        requireAmpResponseSourceOrigin: true,
+      }).onFirstCall().returns(Promise.resolve(mockResponse));
+      return createAdTestingIframePromise().then(fixture => {
+        const doc = fixture.doc;
+        const a4aElement = createA4aElement(doc);
+        a4aElement.setAttribute('width', 200);
+        a4aElement.setAttribute('height', 50);
+        a4aElement.setAttribute('type', 'adsense');
+        const a4a = new MockA4AImpl(a4aElement);
+	const sendXhrRequest_Spy = sandbox.spy(
+            a4a, 'sendXhrRequest_');
+        doc.body.appendChild(a4aElement);
+        a4a.onLayoutMeasure();
+        expect(a4a.adPromise_).to.be.instanceof(Promise);
+        return a4a.adPromise_.then(() => {
+          // Force vsync system to run all queued tasks, so that DOM mutations
+          // are actually completed before testing.
+          a4a.vsync_.runScheduledTasks_();
+	  expect(sendXhrRequest_Spy.calledOnce,
+		 'sendXhrRequest_Spy called exactly once').to.be.true;
+        });
+      });
+    });
+
+   it('should handle network error', () => {
+      viewerWhenVisibleMock.onFirstCall().returns(Promise.resolve());
+      xhrMock.withArgs('https://test.location.org/ad/012345?args', {
+        mode: 'cors',
+        method: 'GET',
+        credentials: 'include',
+        requireAmpResponseSourceOrigin: true,
+      }).returns(Promise.reject("Network Error"));
+      return createAdTestingIframePromise().then(fixture => {
+        const doc = fixture.doc;
+        const a4aElement = createA4aElement(doc);
+        a4aElement.setAttribute('width', 200);
+        a4aElement.setAttribute('height', 50);
+        a4aElement.setAttribute('type', 'adsense');
+	const a4a = new MockA4AImpl(a4aElement);
+	const sendXhrRequest_Spy = sandbox.spy(
+            a4a, 'sendXhrRequest_');
+        doc.body.appendChild(a4aElement);
+        a4a.onLayoutMeasure();
+        expect(a4a.adPromise_).to.be.instanceof(Promise);
+        return a4a.adPromise_.then(() => {
+          // Force vsync system to run all queued tasks, so that DOM mutations
+          // are actually completed before testing.
+          a4a.vsync_.runScheduledTasks_();
+	  expect(sendXhrRequest_Spy.calledOnce,
+		 'sendXhrRequest_Spy called exactly once').to.be.true;
+	  expect(sendXhrRequest_Spy.returnValues.length).to.equal(1);
+	  return sendXhrRequest_Spy.returnValues[0].then(returnedValue => {
+	    expect(returnedValue).to.be.null;
```

Comment:
I'm not clear what exactly is being tested here.  You're forcing the XHR to error out (at least, up to the open issue about error messages from unhandled promise rejections).  You check that `sendXhrRequest` is called, but not how it handles the error, I think?  You should check that it rejects or throws or whatever and, ideally, what the rest of the promise chain does with that result.